java/src/main/java/ai/rapids/cudf/MemoryListener.java

```diff
@@ -0,0 +1,147 @@
+ * we know exactly what it takes, but in others it is really just a guess and it may guess too much
+ * or too little memory. Typically this is used to try and tell if we are getting close to memory
+ * limits and possibly release cached data on the GPU to reduce the amount of memory being used.
+ * In those cases a prediction should be
+ * treated like a memory reservation that is deducted from when an allocation happens until it is
+ * used up.
+ * <br />
+ * Please note that if your interactions with cuDF are threaded the MemoryListener will be called
+ * from each of those threads too. In this case each allocation or endPrediction call is associated
+ * with the previous prediction call on the same thread.
+ */
+public abstract class MemoryListener {
+  /**
+   * A prediction about how much memory is about to be used for an operation. Predictions will not
+   * nest but each thread is independent of other threads.
+   * @param amount the number of bytes predicted to be used.
+   * @param note a string that can be used for debugging to help keep track of what is making the
+   *             prediction.
+   */
+  public abstract void prediction(long amount, String note);
+
+  /**
+   * An actual memory allocation has occurred.
+   * @param amount the number of bytes actually allocated.
+   * @param id a unique number that can be used for debugging. The same number will be sent to
+   *           deallocate.
+   */
+  public abstract void allocation(long amount, long id);
+
+  /**
+   * Indicates that the previous prediction on this thread is no longer in force. Predictions should
+   * never nest.
+   * @param note a string that can be used for debugging.  This should be the same string that was
+   *             passed into prediction.
+   */
+  public abstract void endPrediction(String note);
+
+  /**
+   * Memory was actually deallocated.
+   * @param amount the number of bytes released and can be reused.
+   * @param id a unique number that can be used for debugging.  The same number would have been sent
+   *           to allocate.
+   */
+  public abstract void deallocation(long amount, long id);
+
+  /**
+   * Holds the set of all listeners. An <code>AtomicReference&lt;HashSet&gt;</code> is used here to
+   * optimize the common read path over the less common write path (add/remove listeners). Other
+   * data structures like a ConcurrentHashMap still involve a lock on the read path where as this
+   * eliminates any locking on reads.
+   */
+  private static final AtomicReference<HashSet<MemoryListener>> listeners =
+      new AtomicReference<>(new HashSet<>());
+
+  /**
+   * Register a memory listener. If the memory listener is already registered it will not install
+   * duplicates. This should be done before any column operations happen or you risk missing those
+   * operations.
+   * @param listener the listener to start sending events to.
+   */
+  public static void registerDeviceListener(final MemoryListener listener) {
```

Comment:
should this be `registerMemoryListener`?

Reply:
No because I wanted to potentially reuse the same interface in the future for pinned host memory or just regular host memory